### PR TITLE
Restructure Lite examples and add GAN

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

- Moves the MNIST example to a subfolder
- Adds the dcgan code from https://github.com/Lightning-AI/LAI-GAN-examples

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: examples/fabric/image_classifier/train_torch.py [hunk 1]
```diff
@@ -15,18 +15,44 @@
...
+import torch.nn as nn
 import torch.nn.functional as F
 import torch.optim as optim
 import torchvision.transforms as T
-from models import Net
 from torch.optim.lr_scheduler import StepLR
 from torchvision.datasets import MNIST
 
-DATASETS_PATH = path.join(path.dirname(__file__), "..", "..", "Datasets")
+DATASETS_PATH = path.join(path.dirname(__file__), "../..", "..", "Datasets")
```
Comment (carmocca) [inline examples/fabric/image_classifier/train_torch.py:25]:
```suggestion
DATASETS_PATH = path.join(path.dirname(__file__), "..", "..", "..", "Datasets")
```

File: examples/fabric/image_classifier/train_torch.py [hunk 2]
```diff
@@ -15,18 +15,44 @@
...
 
 
 # Credit to the PyTorch team
 # Taken from https://github.com/pytorch/examples/blob/master/mnist/main.py and slightly adapted.
+class Net(nn.Module):
```
Comment (carmocca) [inline examples/fabric/image_classifier/train_torch.py:30]: Can we import the version in `src/pytorch_lightning/demos/boring_classes.py`?

File: examples/fabric/image_classifier/train_fabric.py [hunk 3]
```diff
@@ -32,18 +32,44 @@
...
 from torchvision.datasets import MNIST
 
 from lightning.fabric import Fabric  # import Fabric
 from lightning.fabric import seed_everything
 
-DATASETS_PATH = path.join(path.dirname(__file__), "..", "..", "Datasets")
+DATASETS_PATH = path.join(path.dirname(__file__), "../..", "..", "Datasets")
+
+
+class Net(nn.Module):
```
Comment (carmocca) [inline examples/fabric/image_classifier/train_fabric.py:49]: Same suggestion

File: examples/fabric/image_classifier/train_fabric.py [hunk 4]
```diff
@@ -32,18 +32,44 @@
...
-from models import Net
 from torch.optim.lr_scheduler import StepLR
 from torchmetrics.classification import Accuracy
 from torchvision.datasets import MNIST
 
 from lightning.fabric import Fabric  # import Fabric
 from lightning.fabric import seed_everything
 
-DATASETS_PATH = path.join(path.dirname(__file__), "..", "..", "Datasets")
+DATASETS_PATH = path.join(path.dirname(__file__), "../..", "..", "Datasets")
```
Comment (carmocca) [inline examples/fabric/image_classifier/train_fabric.py:46]:
```suggestion
DATASETS_PATH = path.join(path.dirname(__file__), "..", "..", "..", "Datasets")
```

File: examples/fabric/dcgan/train_torch.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
+"""
+DCGAN - Raw PyTorch Implementation
+
+Code adapted from the official PyTorch DCGAN tutorial:
```
Comment (carmocca) [inline examples/fabric/dcgan/train_torch.py:4]: What exactly did you adapt from this file? Or are there no changes from the reference?

File: examples/README.md [hunk 6]
```diff
@@ -16,8 +16,10 @@ ______________________________________________________________________
 We show how to accelerate your PyTorch code with [Lightning Fabric](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_fabric.html) with minimal code changes.
 You stay in full control of the training loop.
 
-- [MNIST with vanilla PyTorch](fabric/image_classifier_1_pytorch.py)
-- [MNIST with Lightning Fabric](fabric/image_classifier_2_fabric.py)
+- [MNIST with vanilla PyTorch](fabric/image_classifier/train_torch.py)
+- [MNIST with Lightning Fabric](fabric/image_classifier/train_fabric.py)
+- [DCGAN vanilla PyTorch](fabric/dcgan/train_torch.py)
+- [DCGAN with Lightning Fabric](fabric/dcgan/train_fabric.py)
```
Comment (carmocca) [inline examples/README.md:22]: I would add suggestions here to visualize the differences.

For example by running `sdiff examples/fabric/dcgan/train_{torch,fabric}.py`

**Review by carmocca — APPROVED** [review]

File: examples/fabric/image_classifier/train_torch.py (hunk 2 above)
Comment (awaelchli) [reply examples/fabric/image_classifier/train_torch.py:30]: I want to keep the file fully self-contained. This is a very trivial net, and we don't expect there to be major changes to it

File: examples/fabric/dcgan/train_torch.py (hunk 5 above)
Comment (awaelchli) [reply examples/fabric/dcgan/train_torch.py:4]: I don't remember. Took the code, formatted it. 
This is just to say where it came from. If we ever make changes, the comment covers it.

File: examples/fabric/dcgan/train_fabric.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+num_epochs = 5
+# Learning rate for optimizers
+lr = 0.0002
+# Beta1 hyperparameter for Adam optimizers
+beta1 = 0.5
+# Number of GPUs to use
+num_gpus = 1
+
+
+def main():
```
Comment (Borda) [inline examples/fabric/dcgan/train_fabric.py:48]: just thinking to make all the default/constants above as arguments here so that with a simple use of [`Fire`](https://github.com/google/python-fire) for CLI creation all these parameters can be exposed in CLI :rabbit:

File: examples/fabric/dcgan/train_fabric.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    losses_d = []
+    iteration = 0
+
+    # Training loop
+    for epoch in range(num_epochs):
+        for i, data in enumerate(dataloader, 0):
+            # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
+            # (a) Train with all-real batch
+            discriminator.zero_grad()
+            real_cpu = data[0]
```
Comment (Borda) [inline examples/fabric/dcgan/train_fabric.py:126]: why CPU? it is confusing name

Comment (Borda) [reply examples/fabric/dcgan/train_fabric.py:126]:
```suggestion
            real_spl = data[0]
```

File: examples/fabric/dcgan/train_torch.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    optimizer_g = optim.Adam(generator.parameters(), lr=lr, betas=(beta1, 0.999))
+
+    # Lists to keep track of progress
+    losses_g = []
+    losses_d = []
+    iteration = 0
+
+    # Training loop
+    for epoch in range(num_epochs):
+        for i, data in enumerate(dataloader, 0):
```
Comment (Borda) [inline examples/fabric/dcgan/train_torch.py:128]: tbh, it is confusing, looking up and down and the two training loops with and without fabric look very the same so as a user it not clear what is the fabric benefit... :/

**Review by Borda — APPROVED** [review]

File: examples/fabric/dcgan/train_fabric.py (hunk 7 above)
Comment (awaelchli) [reply examples/fabric/dcgan/train_fabric.py:48]: This is the standard dcgan example from pytorch, I do not plan to refactor it in such ways.

File: examples/fabric/dcgan/train_fabric.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+            b_size = real_cpu.size(0)
+            label = torch.full((b_size,), real_label, dtype=torch.float, device=fabric.device)
+            # Forward pass real batch through D
+            output = discriminator(real_cpu).view(-1)
+            # Calculate loss on all-real batch
+            err_d_real = criterion(output, label)
```
Comment (awaelchli) [inline examples/fabric/dcgan/train_fabric.py:132]:
```suggestion
            real = data[0]
            b_size = real.size(0)
            label = torch.full((b_size,), real_label, dtype=torch.float, device=fabric.device)
            # Forward pass real batch through D
            output = discriminator(real).view(-1)
            # Calculate loss on all-real batch
            err_d_real = criterion(output, label)
```

File: examples/fabric/dcgan/train_fabric.py (hunk 8 above)
Comment (awaelchli) [reply examples/fabric/dcgan/train_fabric.py:126]: because in train_torch.py, they moved it to the device later on.

File: examples/fabric/dcgan/train_torch.py (hunk 9 above)
Comment (awaelchli) [reply examples/fabric/dcgan/train_torch.py:128]: Yes, Fabric does not significantly change the code. There is not much boilerplate code in this example, so Fabric does very little for the user here. It is however possible to scale the code more easily, in ways that would otherwise require more boilerplate code to be added to the existing raw pytorch example.
